### PR TITLE
Remove SourceDetection rename source_detection to source_catalog

### PR DESCRIPTION
**Pull request by braingram**
This PR:
- removes `SourceDetectionStep`
- update romancal to accomdate changes in https://github.com/spacetelescope/rad/pull/513 which renamed
  - `meta.source_detection` to `meta.source_catalog`
  - `meta.cal_step.source_detection` to `meta.cal_step.souce_catalog`
  - `SourceCatalog` to `ImageSourceCatalog` (to deal with schema name conflicts with the above renaming and to make this agree with `MosaicSourceCatalog`)

As the above renaming breaks reading current files a new set of regtest data was generated:
`roman-pipeline/dev_20241127/`
all tests pass locally.
This PR also makes several changes to `make_regtest_data.sh` to deal with a few issues. I will call them out in comments.

Regtests (with existing files showing many expected failures): https://github.com/spacetelescope/RegressionTests/actions/runs/11942934325

Running regtests with new truth files: https://github.com/spacetelescope/RegressionTests/actions/runs/12144666599

## Tasks
- [ ] **request a review from someone specific**, to avoid making the maintainers review every PR
- [ ] add a build milestone, i.e. `24Q4_B15` (use the [latest build](https://github.com/spacetelescope/romancal/milestones) if not sure)
- [ ] Does this PR change user-facing code / API? (if not, label with `no-changelog-entry-needed`)
  - [ ] write news fragment(s) in `changes/`: `echo "changed something" > changes/<PR#>.<changetype>.rst` (see below for change types)
  - [ ] update or add relevant tests
  - [ ] update relevant docstrings and / or `docs/` page
  - [ ] [start a regression test](https://github.com/spacetelescope/RegressionTests/actions/workflows/romancal.yml) and include a link to the running job ([click here for instructions](https://github.com/spacetelescope/RegressionTests/blob/main/docs/running_regression_tests.md))
    - [ ] Do truth files need to be updated ("okified")?
      - [ ] **after the reviewer has approved these changes**, run `okify_regtests` to update the truth files
- [ ] if a JIRA ticket exists, [make sure it is resolved properly](https://github.com/spacetelescope/romancal/wiki/How-to-resolve-JIRA-issues)

<details><summary>news fragment change types...</summary>

  - ``changes/<PR#>.general.rst``: infrastructure or miscellaneous change
  - ``changes/<PR#>.docs.rst``
  - ``changes/<PR#>.stpipe.rst``
  - ``changes/<PR#>.associations.rst``
  - ``changes/<PR#>.scripts.rst``
  - ``changes/<PR#>.mosaic_pipeline.rst``
  - ``changes/<PR#>.patch_match.rst``

  ## steps
  - ``changes/<PR#>.dq_init.rst``
  - ``changes/<PR#>.saturation.rst``
  - ``changes/<PR#>.refpix.rst``
  - ``changes/<PR#>.linearity.rst``
  - ``changes/<PR#>.dark_current.rst``
  - ``changes/<PR#>.jump_detection.rst``
  - ``changes/<PR#>.ramp_fitting.rst``
  - ``changes/<PR#>.assign_wcs.rst``
  - ``changes/<PR#>.flatfield.rst``
  - ``changes/<PR#>.photom.rst``
  - ``changes/<PR#>.flux.rst``
  - ``changes/<PR#>.source_detection.rst``
  - ``changes/<PR#>.tweakreg.rst``
  - ``changes/<PR#>.skymatch.rst``
  - ``changes/<PR#>.outlier_detection.rst``
  - ``changes/<PR#>.resample.rst``
  - ``changes/<PR#>.source_catalog.rst``
</details>

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -23,8 +23,8 @@ dependencies = [
     "photutils >=1.13.0",
     "pyparsing >=2.4.7",
     "requests >=2.26",
-    # "roman_datamodels>=0.22.0,<0.23.0",
-    "roman_datamodels @ git+https://github.com/spacetelescope/roman_datamodels.git",
+    #"roman_datamodels>=0.22.0,<0.23.0",
+    "roman_datamodels @ git+https://github.com/braingram/roman_datamodels.git@source_catalog",
```
Comment (braingram) [inline pyproject.toml:27]: Temporarily pointing to the source branch for https://github.com/spacetelescope/roman_datamodels/pull/428 to pick up required changes.

File: romancal/scripts/make_regtestdata.sh [hunk 2]
```diff
@@ -36,7 +53,7 @@ do
     cp ${fn}_*.asdf $outdir/roman-pipeline/dev/truth/WFI/$dirname/
 
     # uncal file doesn't belong in truth.
-    rm $outdir/roman-pipeline/dev/truth/WFI/image/${fn}_uncal.asdf
+    rm $outdir/roman-pipeline/dev/truth/WFI/${dirname}/${fn}_uncal.asdf
```
Comment (braingram) [inline romancal/scripts/make_regtestdata.sh:56]: Minor bugfix. The loop processing the `grism` tries to delete the `image` file (and the `grism` uncal isn't deleted).

File: romancal/scripts/make_regtestdata.sh [hunk 3]
```diff
@@ -63,11 +80,11 @@ cp r0000101001001001001_0002_wfi10_cal.asdf $outdir/roman-pipeline/dev/WFI/image
 
 # resample regtest; needs r0000101001001001001_000{1,2}_wfi01_cal.asdf
 # builds the appropriate asn file and calls strun with it
-echo "Creating regtest files for resample..."
-asn_from_list r0000101001001001001_0001_wfi01_cal.asdf r0000101001001001001_0002_wfi01_cal.asdf r0000101001001001001_0003_wfi01_cal.asdf -o L3_mosaic_asn.json --product-name mosaic
-strun romancal.step.ResampleStep L3_mosaic_asn.json --rotation=0 --output_file=mosaic.asdf
-cp L3_mosaic_asn.json $outdir/roman-pipeline/dev/WFI/image/
-cp mosaic_resamplestep.asdf $outdir/roman-pipeline/dev/truth/WFI/image/
+#echo "Creating regtest files for resample..."
+#asn_from_list r0000101001001001001_0001_wfi01_cal.asdf r0000101001001001001_0002_wfi01_cal.asdf r0000101001001001001_0003_wfi01_cal.asdf -o L3_mosaic_asn.json --product-name mosaic
```
Comment (braingram) [inline romancal/scripts/make_regtestdata.sh:84]: `L3_mosaic_asn.json` is generated twice in this file on main (and each time the association is different). On artifactory the current `L3_mosaic_asn.json` has a target (`r274dp63x31y81`) which better matches the other place where `L3_mosaic_asn.json` is generated so this portion of the script was commented out (let me know if I should just delete these lines) and the other updated to generate the required files using a single `L3_mosaic_asn.json`.

File: romancal/scripts/make_regtestdata.sh [hunk 4]
```diff
@@ -120,10 +137,8 @@ cp Test_linearity.asdf $outdir/roman-pipeline/dev/truth/WFI/image/
 
 # we have a test that runs the flat field step directly on an _L1_ spectroscopic
 # file and verifies that it gets skipped.
-# I don't really understand that but we can duplicate it for now.
-# FIXME: is the uncal input for test_flat_field_grism_step() correct?
```
Comment (braingram) [inline romancal/scripts/make_regtestdata.sh:124]: This is fixed by using the `_assignwcs` file.

File: romancal/scripts/make_regtestdata.sh [hunk 5]
```diff
@@ -206,19 +231,15 @@ cp ${l3name}_cat.asdf $outdir/roman-pipeline/dev/truth/WFI/image/
 strun romancal.step.SourceCatalogStep r0000101001001001001_0001_wfi01_cal.asdf
 cp r0000101001001001001_0001_wfi01_cat.asdf $outdir/roman-pipeline/dev/truth/WFI/image/
 
-# L3 on skycell
-l3name="r0099101001001001001_r274dp63x31y81_prompt_F158"
-asn_from_list --product-name=$l3name r0000101001001001001_0001_wfi01_cal.asdf r0000101001001001001_0002_wfi01_cal.asdf r0000101001001001001_0003_wfi01_cal.asdf -o L3_m1_asn.json
-strun roman_mos L3_m1_asn.json
```
Comment (braingram) [inline romancal/scripts/make_regtestdata.sh:212]: `L3_m1_asn.json` doesn't exist on artifactory.

File: romancal/scripts/make_regtestdata.sh [hunk 6]
```diff
@@ -206,19 +231,15 @@ cp ${l3name}_cat.asdf $outdir/roman-pipeline/dev/truth/WFI/image/
...
-
 # L3 skycell catalog
 strun romancal.step.SourceCatalogStep ${l3name}_i2d.asdf
 cp ${l3name}_cat.asdf $outdir/roman-pipeline/dev/truth/WFI/image/
 
 # multiband catalog
 asn_from_list --product-name=${l3name}_mbcat ${l3name}_i2d.asdf -o L3_skycell_mbcat_asn.json
 strun romancal.step.MultibandCatalogStep L3_skycell_mbcat_asn.json --deblend True
+cp L3_skycell_mbcat_asn.json $outdir/roman-pipeline/dev/WFI/image/
+cp ${l3name}_mbcat_cat.asdf $outdir/roman-pipeline/dev/truth/WFI/image/
```
Comment (braingram) [inline romancal/scripts/make_regtestdata.sh:242]: These files weren't being copied to the output directory.

Comment (schlafly) [reply romancal/scripts/make_regtestdata.sh:242]: Good catch, thanks.

File: romancal/scripts/make_regtestdata.sh (hunk 3 above)
Comment (schlafly) [reply romancal/scripts/make_regtestdata.sh:84]: Go ahead and delete these lines.

Comment (braingram) [reply romancal/scripts/make_regtestdata.sh]: Thanks! Removed in [6050049](https://github.com/spacetelescope/romancal/pull/1533/commits/6050049d368843dacbe5a5e3827011d80b7dc232)

**Review by schlafly — APPROVED** [review]
LGTM, thanks!

**Pull request merged**